### PR TITLE
Metrics

### PR DESCRIPTION
**Pull request by dehall**
Adds optional metric collection on state transition data, intended to be useful for debugging. This is largely based on the `metrics` branch on ruby synthea which was never merged:  https://github.com/synthetichealth/synthea/compare/metrics

This data is only collected if the config setting `generate.track_detailed_transition_metrics` is enabled. (disabled by default)

Sample output (truncated because it gets very long):

```
METABOLIC SYNDROME DISEASE PROGRESSION MODULE
Age_Guard:
 Total times entered: 123
 Population that ever hit this state: 123 (100.00%)
 Average # of hits per total population: 100.00
 Average # of hits per person that ever hit state: 100.00
 Population currently in state: 22 (17.89%)
 Total duration: 2010 years 6 months
 Average duration per time entered: 16 years 4 months
 Average duration per person that ever entered state: 16 years 4 months
 Transitioned to:
 --> Chance_to_Onset_Hypertension : 101 = 100.00%

Initial_Kidney_Health:
 Total times entered: 123
 Population that ever hit this state: 123 (100.00%)
 Average # of hits per total population: 100.00
 Average # of hits per person that ever hit state: 100.00
 Population currently in state: 0 (0.00%)
 Transitioned to:
 --> Initial_Eye_Health : 123 = 100.00%

Initial:
 Total times entered: 123
 Population that ever hit this state: 123 (100.00%)
 Average # of hits per total population: 100.00
 Average # of hits per person that ever hit state: 100.00
 Population currently in state: 0 (0.00%)
 Transitioned to:
 --> Initial_Kidney_Health : 123 = 100.00%

No_Diabetes:
 Total times entered: 77
 Population that ever hit this state: 77 (62.60%)
 Average # of hits per total population: 62.60
 Average # of hits per person that ever hit state: 100.00
 Population currently in state: 77 (62.60%)

Diabetes_Prevalence:
 Total times entered: 101
 Population that ever hit this state: 101 (82.11%)
 Average # of hits per total population: 82.11
 Average # of hits per person that ever hit state: 100.00
 Population currently in state: 0 (0.00%)
 Transitioned to:
 --> No_Diabetes : 47 = 46.53%
 --> Eventual_Prediabetes : 44 = 43.56%
 --> Eventual_Diabetes : 10 = 9.90%

Onset_Prediabetes:
 Total times entered: 30
 Population that ever hit this state: 30 (24.39%)
 Average # of hits per total population: 24.39
 Average # of hits per person that ever hit state: 100.00
 Population currently in state: 0 (0.00%)
 Transitioned to:
 --> No_Diabetes : 30 = 100.00%

Eventual_Prediabetes:
 Total times entered: 44
 Population that ever hit this state: 44 (35.77%)
 Average # of hits per total population: 35.77
 Average # of hits per person that ever hit state: 100.00
 Population currently in state: 14 (11.38%)
 Total duration: 696 years 1 months
 Average duration per time entered: 15 years 9 months
 Average duration per person that ever entered state: 15 years 9 months
 Transitioned to:
 --> Onset_Prediabetes : 30 = 100.00%

Diabetes_Progression:
 Total times entered: 2129
 Population that ever hit this state: 8 (6.50%)
 Average # of hits per total population: 1730.89
 Average # of hits per person that ever hit state: 26612.50
 Population currently in state: 8 (6.50%)
 Total duration: 203 years 9 months
 Average duration per time entered: 1 months 4 days
 Average duration per person that ever entered state: 25 years 5 months
 Transitioned to:
 --> Nephropathy_Progression : 2121 = 100.00%
```

### CODE REVIEW

**Comment by jawalonoski** [comment]
The "about" phrases seem off. For example,

`5798.07 years (About 5798 years and 7 months)`

File: src/main/java/org/mitre/synthea/helpers/TransitionMetrics.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+  private synchronized Metric getMetric(String moduleName, String stateName) {
+    Metric metric = metrics.get(moduleName, stateName);
+
+    if (metric == null) {
+      metric = new Metric();
+      metrics.put(moduleName, stateName, metric);
+    }
+
+    return metric;
```
Comment (jawalonoski) [inline src/main/java/org/mitre/synthea/helpers/TransitionMetrics.java]: Hmm... `getMetric(...)` is synchronized, but the returned `Metric` is not. These then get incremented in a non-thread safe way.

File: src/main/java/org/mitre/synthea/helpers/TransitionMetrics.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+   */
+  private static class Metric {
+    int entered; // number of times this state was entered
+    long duration; // total length of time people sat in this state
+    public int population; // number of people that ever his this state
+    int current; // number of people that are "currently" in that state
+    Map<String, Integer> destinations = new HashMap<>(); // key: state that this state transitioned
+                                                         // to, value: number of times
+
+    void incrementDestination(String destination) {
```
Comment (jawalonoski) [inline src/main/java/org/mitre/synthea/helpers/TransitionMetrics.java]: You should probably make this thread safe.

File: src/main/java/org/mitre/synthea/helpers/TransitionMetrics.java [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    // note that this is especially helpful because ints can be passed in without explicit casting
+    // and if you want to get a double from integer division you have to cast the input items
+    return String.format("%.2f", (100.0 * num / denom));
+  }
+
+  /**
+   * Helper class to track the metrics of a single State.
+   */
+  private static class Metric {
+    int entered; // number of times this state was entered
```
Comment (jawalonoski) [inline src/main/java/org/mitre/synthea/helpers/TransitionMetrics.java]: comments should be javadoc before the declaration.

For example:
```java
/** number of times this state was entered. */
int entered;
```

**Comment by dehall** [comment]
@jawalonoski All comments addressed in the latest version

**Comment by jawalonoski** [comment]
Looks good. Fix conflicts then we'll merge.

**Pull request merged**